test/dist_autograd_test.py

```diff
@@ -688,6 +732,139 @@ def test_backward_simple(self):
+        local_grads = None
+        t1 = torch.ones((3, 3), requires_grad=True)
+        t2 = torch.zeros((3, 3), requires_grad=True)
+
+        local_ret = torch.add(t1, t2)
+        local_ret.sum().backward()
+        with dist_autograd.context() as context_id:
+            rref_t1 = rpc.remote(
+                rref_owner,
+                _torch_ones,
+                args=((3, 3),),
+                kwargs={"requires_grad": True}
+            )
+
+            if callee == rref_owner:
+                rref = rpc.remote(callee, my_rref_add, args=(rref_t1, t2))
+            else:
+                rref = rpc.remote(
+                    callee,
+                    my_nested_rref_add,
+                    args=(rref_owner, rref_t1, t2)
+                )
+            ret = rref.to_here().wait()
+            dist_autograd.backward([ret.sum()])
+
+            # verify grads on caller
+            grads = dist_autograd.get_gradients(context_id)
+            self.assertIn(t2, grads)
+            self.assertEqual(grads[t2], t2.grad)
+
+            # verify grads on rref owner
+            self.assertTrue(
+                rpc.rpc_sync(
+                    rref_owner,
+                    _compare_owner_value,
+                    args=(context_id, rref_t1, t1.grad)
+                )
+            )
+
+    @dist_init
+    def test_backward_rref(self):
+        callee = "worker{}".format(self._next_rank())
+        rref_owner = callee
+        self._test_backward_rref(callee, rref_owner)
+
+    @dist_init
+    def test_backward_rref_multi(self):
+        if self.rank > 0:
+            callee = "worker0"
+            rref_owner = callee
+            self._test_backward_rref(callee, rref_owner)
+
+    @dist_init
+    def test_backward_rref_nested(self):
+        callee = "worker{}".format((self.rank + 1) % self.world_size)
+        rref_owner = "worker{}".format((self.rank + 2) % self.world_size)
+        self._test_backward_rref(callee, rref_owner)
+
+    @dist_init
+    def test_backward_rref_multi_user(self):
+        local_grads = None
+        t1 = torch.ones((3, 3), requires_grad=True)
+        t2 = torch.zeros((3, 3), requires_grad=True)
+
+        local_ret = torch.add(t1, t2)
+        local_ret.sum().backward()
+
+        # create rref on self
+        # TODO: simplify this once we support rpc to self
+        self_name = "worker{}".format(self.rank)
+        rref_t1 = rpc.rpc_sync(
+            "worker{}".format(self._next_rank()),
+            _create_ones_rref_on,
+            args=(self_name, (3, 3))
+        )
+
+        # kick off forward and backward pass on three other workers (trainers)
+        rank_diffs = [1, 2, 3]
+        futures = []
+        for rank_diff in rank_diffs:
+            futures.append(rpc.rpc_async(
+                "worker{}".format((self.rank + rank_diff) % self.world_size),
+                _run_trainer,
+                args=(rref_t1, t2, t2.grad, self_name, rank_diff)
```

Comment:
Not sure I follow the test completely. It looks like the `ps` is different for each node (the ps is self.rank). Shouldn't make rank 0 the ps and ranks 1, 2 and 3 be trainers?

Reply:
This is testing 4 ps and 12 trainers concurrently. Each process x will serve as a ps and driver which will kick off 3 trainers that only talks to ps x. So we have

ps = rank0, trainer = rank1, 2, 3
ps = rank1, trainer = rank 2, 3, 4
....

Let me add some comments here to explain.